rfc-091-sharing-assets.md

```diff
@@ -0,0 +1,43 @@
+# Sharing assets
+
+## Summary
+
+We need to share assets to make sure users only have to download a thing once. To do this, we'll upload each application's assets to the same location.
+
+## Problem
+
+In [RFC84 we proposed replacing Static with a gem](https://github.com/alphagov/govuk-rfcs/pull/84). One of the architectural challenges with this approach is that by moving assets (like CSS, JS and images) into the gem, the assets won't be shared for users. This means that users visiting GOV.UK would have to download assets multiple times, depending on what app renders the page.
+
+## Proposal
+
+We upload the assets to a central location. For example, if both `collections` and `government-frontend` have a file named `foo.css`, they will upload the file to `assets.publishing.service.gov.uk/shared/foo.css`. Because [Rails fingerprints assets](http://guides.rubyonrails.org/asset_pipeline.html#what-is-fingerprinting-and-why-should-i-care-questionmark), the asset actually uploaded will be `assets.publishing.service.gov.uk/shared/foo-8d811b8c3badbc0b0e2f6e25d3660a96cc0cca7993e6f32e98785f205fc40907.css`, unless the files are different, in which case they'll have different fingerprints and won't overwrite.
+
+How to achieve this:
+
+* In govuk-app-deployment, we configure capistrano to precompile the assets and then upload the resulting files to a central location.
+* Proxy assets.publishing.service.gov.uk/shared to this location
+* Point applications to use this URL as asset host so that it's used in production
+
+Upsides
+
+* We remove Nginx config to proxy to individual applications, perhaps removing some hops
+* Whenever there's an asset to share, it will be shared automatically. If an application copies an image from another app, on production they'll have the same URL.
+* Each application ensures that its own assets are present on the shared bucket.
+* Applications don't serve files anymore.
+
+Downsides
+
+* We'll end up with a directory full of assets which are of unknown origin.
+
+## Alternatives
+
+### 1. Do nothing
+
+The simplest solution is to not share assets between applications. This would have significant impact on the user, especially if the font will be distributed in the gem.
```

Comment:
A problem that have in terms of measuring the benefit of any of this is that we don't know the impact to users of any of these sorts of decisions.  I wrote [a card for our backlog](https://trello.com/c/wPcGWJbi/386-set-up-performance-testing-metrics-for-govuk-pages) that suggests running [some web page tests](https://www.webpagetest.org/) across various sorts of pages and uploading metrics to graphite, but I don't know when we're going to get to it.

Reply:
I can help put these measurements together, let me know, I'll try to find some time to do help.

Reply:
While it's true that we don't know exactly how much impact this will have, it's certain that if we were to move all assets to per-app distribution there would be more things to download.

If we look at someone who goes from the homepage to a browse page:

- https://www.gov.uk without cache is 458KB
- https://www.gov.uk/browse/disabilities without cache is 351KB
- https://www.gov.uk/browse/disabilities after visiting the homepage is 80KB

So at the very least this journey would mean 271KB more data traffic for this particular journey. Most of it is the font of course (~260KB) but the principle probably stands.